### DISCUSSION TITLE
Igor Pro 6 support?

### DISCUSSION THREAD

**Discussion by nsulmol** [AllenInstitute/ZeroMQ-XOP#50] [Q&A]
Hello!

I am looking to compile ZeroMQ-XOP for Igor Pro 6, as the software I am working with is stuck on it (Asylum Research). I have taken a quick look at the mentioned differences (from a programmer's perspective) between 6 and 7+: https://www.wavemetrics.com/products/xoptoolkit/versions

I think I have most of the major changes handled in a fork (though this will be more obvious once I try to compile). My goal was to purchase the XOP toolkit and iterate until I have something functional.

Before purchasing it, I wanted to confirm there isn't any *obvious* limitation why it is not possible to get ZeroMQ-XOP running with this older version of Igor Pro 6. Can you think of any?

Thank you,
Nick

**Comment by t-b** [comment]
> Before purchasing it, I wanted to confirm there isn't any obvious limitation why it is not possible to get ZeroMQ-XOP running with this older version of Igor Pro 6. Can you think of any?

No, I don't think we use any IP7+ features in the XOP itself. The tests will likely not run in IP6 as the builtin JSON operation is missing.

**Reply by nsulmol** [reply]
Hello t-b,

Thank you for the quick reply! I will purchase and test tomorrow.

Out of curiosity: do you mean that the tests themselves use a builtin JSON operation? With a quick fzf search in tests, I found references to JSONSimple, perhaps that is what you are referring to? If so, are these part of the JSON XOP you developed at byte physics (presumably by you)?

Also, am I correct in understanding that RunTests.pxp is effectively an Igor 'script' for running all the tests in a particular order, with potentially predefined inputs/outputs? I am trying to understand this versus the individual test ipf files.

Excuse all the questions, I will be next to a Windows distribution with Igor installed tomorrow morning so perhaps these questions are silly.

Thank you,
Nick

**Reply by t-b** [reply]
> do you mean that the tests themselves use a builtin JSON operation? With a quick fzf search in tests, I found references to JSONSimple, perhaps that is what you are referring to?

Yes. Back when the ZeroMQ XOP was developed, there was no JSON XOP. So I asked WM and they told me that there is an undocumented operation JSONSimple for parsing JSON. The JSONSimple operation is used to parse JSON. But it's rather limited and does not support a lot of JSON features. The JSON XOP has full-blown JSON support. But the JSON XOP currently also requires IP8+.

>  If so, are these part of the JSON XOP you developed at byte physics (presumably by you)?

No, JSONSimple is builtin into IP but hidden.

> Also, am I correct in understanding that RunTests.pxp is effectively an Igor 'script' for running all the tests in a particular order, with potentially predefined inputs/outputs? I am trying to understand this versus the individual test ipf files.

Yes kind of. RunTests.pxp just holds some code to run the tests. The tests are written using [igortest](https://github.com/byte-physics/igortest). You can also just run each test ipf file separately. We usually have an experiment which implements a `run` function so that running the tests can be automated, see https://docs.byte-physics.de/igortest/advanced.html#automate-test-runs.

> Excuse all the questions, I will be next to a Windows distribution with Igor installed tomorrow morning so perhaps these questions are silly.

No problem at all. In case you want us to perform the downgrading for ZeroMQ XOP (and JSON XOP?) I'm happy to send you an offer.

**Comment by nsulmol** [comment]
Hi @t-b ,

Thanks for the response! I wanted to update you on this: I have a branch in my fork that compiles with XOP Toolkit 7 and appears to run with Igor Pro 6.

I should note that I tested by using the example C++ project, and it functions as expected. I did not attempt testing with the IUTF, as I had validated that the method JSONSimple does not exist in Igor Pro 6.

I was wondering: would you be interested in a pull request for this? I could do so after the separate 'readme update' request I have sent you.

Of note:
- The changes to the actual code were quite minor. All I did was (a) rename a set of variables so that they fit within the 31 character limit of Igor Pro 6/7, and (b) add an '#ifdef #define' for a preprocessor directive that does not exist.
- I add a cmake option IGOR6_OR_7, which - if true - changes the Igor path to an updated one, and updates the minimum supported Igor version.
- I change the #pragma IgorVersion for the ZeroMQ_interop.ipf file (but not the zmq_main_test.ipf, as we cannot run IUTF with it).

If you think it is worth it, I would update the README to explain this option (as an 'experimental' option), and send you a pull request. I can send you the executables for a proper release, but I'm not sure how you would like to set up the zip folder structure for a release (assuming you wanted to do a release).

Let me know your thoughts!

P.S.: It occurs to me that I do not know the purpose of the launcher created by CreateLauncher.scmake. It stays in the build directory and appears to be purposefully *not* copied on install. Is this tied to the unit tests? I recall reading its purpose, but appear to have forgotten to write it down (and now cannot re-find a reference to it!).

Thanks,
Nick

**Reply by t-b** [reply]
> P.S.: It occurs to me that I do not know the purpose of the launcher created by CreateLauncher.scmake. It stays in the build directory and appears to be purposefully not copied on install. Is this tied to the unit tests? I recall reading its purpose, but appear to have forgotten to write it down (and now cannot re-find a reference to it!).

This is code from https://github.com/caseymcc/CreateLaunchers. It allows to directly run Igor Pro with the XOP in a Visual Studio debug session.

Yes I'm interested in principle in a PR. I just need to think about a way to test the resulting XOP.

**Comment by t-b** [comment]
Which IP6 version are we talking about? 6.36 or something earlier?

**Comment by nsulmol** [comment]
Hi Thomas,

I have a branch I'm almost done with, that I will be sending to you later
today. It might avoid duplicated effort if you hold off - just an FYI.

I am putting the minimum requirement as Igor 6.20 because of reasons I
cannot recall (but are in the documentation somewhere). The system I am
actually using is 6.37.

If you are curious about most of my changes, they are here:
https://github.com/nsulmol/ZeroMQ-XOP/tree/igor6support

**Reply by t-b** [reply]
If you don't need support for < 6.37 let's require 6.37. That is already ancient enough.

**Comment by nsulmol** [comment]
Hi @t-b,

A quick update: most of the changes I wanted to propose in a PR are here: https://github.com/nsulmol/ZeroMQ-XOP/tree/igor6support.

However, I am still missing one thing: pushing new versions of the compiled libs/etc in output. One thing you will see in the CL is that I proposed changing the output directory to contain an 'igor6' and 'igor8' directory (to differentiate for people compiling). This means that the CMakeLists must be updated accordingly.

The issue I have run into is that the src/CMakeLists.txt links to libzmq via the output directory - it assumes that libzmq has been copied there, However, in my current configuration libzmq's files are in output/igor8.

Looking at the cmake files, I do not see any inclusion of the libzmq project. Is it possible that the files that are linked to for libzmq (originally in output/$os/libzmq) are not modified by the compilation process? That would explain what I am seeing. If so, was that intentional?

Thanks,
Nick

**Reply by t-b** [reply]
> Looking at the cmake files, I do not see any inclusion of the libzmq project. Is it possible that the files that are linked to for libzmq (originally in output/$os/libzmq) are not modified by the compilation process? That would explain what I am seeing. If so, was that intentional?

The libzmq files are manually copied into the repo. They are built on-demand by the gitlab CI jobs, e.g. https://github.com/AllenInstitute/ZeroMQ-XOP/blob/d260bf504c970728680861bb1eb227fada719e7c/.gitlab-ci.yml#L37.

**Reply by t-b** [reply]
See also 44e9cd79 (output: Add recompiled libzmq, 2021-03-23).

**Comment by t-b** [comment]
Moved to https://github.com/AllenInstitute/ZeroMQ-XOP/pull/55.
